Marlin/src/api/movement/movement.cpp

```diff
@@ -283,28 +291,106 @@ int moveToEndstop(const Endstop& endstop, float f, float maxTravel) {
+      )) {
+        return -1;
+      }
+
+    // Confirm we triggered
+      if (endstopMonitor.acknowledgeTriggered(endstop)) {
+        logError
+          << F("Unable to move to ")
+          << endstop.name
+          << F(", switch did not trigger")
+          << endl;
+        return -1;
+      }
+
+    return vone->stepper.resyncWithStepCount(axis);
+    
+  } else {
+    const float initial_position = current_position[axis]; //record where we start the move, to prevent
+    float measurements[2] = {initial_position, initial_position};
+    float measurement_initial_position = initial_position;
+    float retractDistance = s_defaultRetractDistance[axis];
+    bool consistent = isVirtual ? !forceConsistency : true; //if we want to check for consistency, set 'consistent' to false initially
+    
+    for (int i = 0; i < maxVirtualEndstopCycles; i++){
+
+      //note our location before movement
+      measurement_initial_position = current_position[axis];
+
+      //perform movement
+      if (relativeRawMoveXYZ(
+        axis == X_AXIS ? travel : 0.0f,
+        axis == Y_AXIS ? travel : 0.0f,
+        axis == Z_AXIS ? travel : 0.0f,
+        clampedSpeed,
+        skipMovementSafetyCheck
+      )) {
+        return -1;
+      }
+
+    // Confirm we triggered
+      if (endstopMonitor.acknowledgeTriggered(endstop)) {
+        logError
+          << F("Unable to move to ")
+          << endstop.name
+          << F(", switch did not trigger")
+          << endl;
+        return -1;
+      }
+
+      if(consistent){ //if we have reached consistency, or if forceConsistency = False, or if this is a physical endstop, we skip all the checks for false triggers
+        // Resync with stepper position
+        return vone->stepper.resyncWithStepCount(axis);
+      } else {
+        vone->stepper.resyncWithStepCount(axis);
+        
+        //record the position of this trigger
+        measurements[i % 2] = current_position[axis];
+
+        //make sure that we cannot retract past our initial start point
+        retractDistance = min(s_defaultRetractDistance[axis], abs(measurements[i % 2] - initial_position));
+
+        if (logging_enabled) {
+          log << F("Measured at switch") << endl;
+          log << F("  initial_position: ") << initial_position << endl;
+          log << F("  measurement: ") << measurements[i % 2] << endl;
+          log << F("  retraction: ") << retractDistance << endl;
+        }
+
+        //If we triggered near this attempt's initial starting position, we could be false triggering and should reject
+        if (abs(measurements[i % 2] - measurement_initial_position) <= (abs(retractDistance)*0.25)){
+          log << F("Measurement matched initial position") << endl;
+          retractFromSwitch(endstop, retractDistance);
+          delay(500);
+          continue;
+        } else if(abs(measurements[0] - measurements[1]) < 0.1) {
+          //have found consistent measurment
+          consistent = true;
```

Comment:
It's unclear what forceConsistancy is doing here. We end with a final movement which may or may not be consistant.

Isn't the logic we want:

1) get a position for the endstop.
2) retract.
3) advance to the endstop again:
then:   **consistancy = (lastMovement === retractDistance)**

Also if we find consistant-- we are not retracting, but we are going back up to the RelativeRawMoveXYZ portion of the for loop. Is that correct? Is there a risk that endstopMonitor.acknowledgeTriggered() won't fire if we don't retract. In which case we would return -1 before we get to if(consistent)